src/meta/Base.ts

```diff
@@ -1,45 +1,49 @@
 import { Destroyable } from '@dojo/core/Destroyable';
 import global from '@dojo/shim/global';
-import Map from '@dojo/shim/Map';
-import { WidgetMetaBase, WidgetMetaProperties, WidgetMetaRequiredNodeCallback } from '../interfaces';
+import Set from '@dojo/shim/Set';
+import { WidgetMetaBase, WidgetMetaProperties, NodeHandlerInterface } from '../interfaces';
 
 export class Base extends Destroyable implements WidgetMetaBase {
 	private _invalidate: () => void;
 	private _invalidating: number;
-	private _requiredNodes: Map<string, ([ WidgetMetaBase, WidgetMetaRequiredNodeCallback ])[]>;
-	protected nodes: Map<string, HTMLElement>;
+	protected nodeHandler: NodeHandlerInterface;
+
+	private _requestedNodeKeys = new Set<string>();
 
 	constructor(properties: WidgetMetaProperties) {
 		super();
 
 		this._invalidate = properties.invalidate;
-		this._requiredNodes = properties.requiredNodes;
-
-		this.nodes = properties.nodes;
+		this.nodeHandler = properties.nodeHandler;
 	}
 
 	public has(key: string): boolean {
-		return this.nodes.has(key);
+		return this.nodeHandler.has(key);
 	}
 
-	protected invalidate(): void {
-		global.cancelAnimationFrame(this._invalidating);
-		this._invalidating = global.requestAnimationFrame(this._invalidate);
-	}
+	protected getNode(key: string): HTMLElement | undefined {
+		const node = this.nodeHandler.get(key);
 
-	protected requireNode(key: string, callback?: WidgetMetaRequiredNodeCallback): void {
-		const node = this.nodes.get(key);
-		if (node) {
-			callback && callback.call(this, node);
-		}
-		else {
-			const callbacks = this._requiredNodes.get(key) || [];
-			callback && callbacks.push([ this, callback ]);
-			this._requiredNodes.set(key, callbacks);
-			if (!callback) {
-				this.invalidate();
+		if (!node) {
+			if (this._requestedNodeKeys.has(key)) {
```

Comment:
Won't this throw an error if it's called twice? And is the assumption that `.has` would always be used to check to see if a node exists and never `.getNode`? Throwing an error as a side-effect of a getter could be unexpected.

Reply:
Good catch re the error @pottedmeat. I'll remove the error and only add to the requested keys if it does not already exist in there. Thanks